services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -433,9 +434,21 @@ public AccessTokenResponseBuilder refreshAccessToken(KeycloakSession session, Ur
             responseBuilder.generateIDToken().generateAccessTokenHash();
         }
 
+        storeRefreshTimingInformation(event, refreshToken, validation.newToken);
+
         return responseBuilder;
     }
 
+    /**
+     * Store information to identify early token refreshes of clients which stress the IAM system.
+     */
+    private void storeRefreshTimingInformation(EventBuilder event, RefreshToken refreshToken, AccessToken newToken) {
+        long expirationAccessToken = newToken.getExp() - newToken.getIat();
+        long ageOfRefreshToken = newToken.getIat() - refreshToken.getIat();
+        event.detail(Details.ACCESS_TOKEN_EXPIRATION_TIME, Long.toString(expirationAccessToken));
+        event.detail(Details.AGE_OF_REFRESH_TOKEN, Long.toString(ageOfRefreshToken));
```

Comment:
Isn't the remaining lifetime of the token more relevant that it's age? The age is relative to the lifetime settings when the token was issued.

Reply:
I assume the refreshed pair of refresh tokens and the access token have the same iat time. Then, if my assumption is correct, we should be able to compute the remaining lifetime of the access token as `(refreshToken.getIat() + expirationAccessToken) - newToken.getIat()`.   

But I agree with @ahus1 here that age is more important because we can identify rogue clients more easily.

Another interesting piece of information may be the percentage of the lifetime of the access token. Something like `(ageOfRefreshToken / expirationAccessToken) * 100`. But it can be computed from the data already in the event so I don't think it is necessary to put it there explicitly.